src/passwd.c

```diff
@@ -266,56 +268,69 @@ static int new_password (const struct passwd *pw)
 		}
 	}
 
-	warned = false;
-	for (i = getdef_num ("PASS_CHANGE_TRIES", 5); i > 0; i--) {
-		cp = agetpass (_("New password: "));
+	if (tflg) {
+		/*
+		 * root is setting the passphrase from stdin
+		 */
+		cp = agetpass_stdin ();
 		if (NULL == cp) {
-			MEMZERO(orig);
-			MEMZERO(pass);
 			return -1;
 		}
-		if (warned && (strcmp (pass, cp) != 0)) {
-			warned = false;
-		}
-		STRTCPY(pass, cp);
+		STRTCPY (pass, cp);
```

Comment:
Hmmmm, I need to come back some day to string hardening, and check what to do when STRTCPY() doesn't fit.

From what I can see, this is silently clamping the password to the first 199 bytes +'\0'.  That doesn't look like a good thing to do.

Reply:
Yes, that's what's happening: https://github.com/shadow-maint/shadow/issues/886

Reply:
Thanks!